# Метрики качества классификации ч.1



### 1.1
Выкачайте тренировочный датасет Titanic - train.csv - с сайта [kaggle](https://www.kaggle.com/c/titanic/data). С помощью функции pd.read_csv() загрузите данные в датафрейм. Выведите первые 20 строк и проанализируйте данные: какие колонки присутствуют (более конкретная информация по ним есть на сайте kaggle), каким образом в них обозначены данные и какие типы данных используются (используйте pandas.dtypes). 

In [1]:
import pandas as pd
import numpy as np

In [2]:
titanic = pd.read_csv('train.csv')
titanic_dataframe = pd.DataFrame(titanic)
titanic_dataframe.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
titanic_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [4]:
titanic_dataframe.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

### 1.2
Проверьте, имеются ли пропущенные значения в колонках, и выведите сумму всех пропущенных значений в каждой из колонок.

In [5]:
titanic_dataframe.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### 1.3
Замените все пропущенные значения колонки Age на медианы в зависимости от пола человека: т.е. если пол человека в строке с пропущенным значением "male", заменяете пропущенное значение возраста на медиану по всем известным возрастам мужчин, и наоборот. Выведите медианы возраста в зависимости от пола. Пропущенные значения колонок Cabin и Embarked замените на U (Unknown).

Примечание: для группировки по полу можно использовать метод df.groupby(), который имеет встроенные агрегатные функции (в т.ч. для вычисления медианы).  Для установки пропущенных значений в датасете воспользуйтесь функцией df.apply().

In [6]:
male_avg = titanic_dataframe[titanic_dataframe['Sex'] == 'male']['Age'].mean()
female_avg = titanic_dataframe[titanic_dataframe['Sex'] == 'female']['Age'].mean()

In [7]:
titanic_dataframe.loc[titanic_dataframe.Sex == 'male', 'Age'] = titanic_dataframe.loc[titanic_dataframe.Sex == 'male', 'Age'].fillna(male_avg)
titanic_dataframe.loc[titanic_dataframe.Sex == 'female', 'Age'] = titanic_dataframe.loc[titanic_dataframe.Sex == 'female', 'Age'].fillna(female_avg)

In [8]:
titanic_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [9]:
titanic_dataframe['Cabin'] = titanic_dataframe['Cabin'].fillna('U')
titanic_dataframe['Embarked'] = titanic_dataframe['Embarked'].fillna('U')

In [10]:
titanic_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        891 non-null    object 
 11  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### 1.4
Выведите возраст пассажиров с PassengerID = [6, 20]. Убедитесь, что заполнены все пропущенные значения (воспользуйтесь функцией df.isnull() ).

In [11]:
titanic_dataframe.loc[titanic_dataframe['PassengerId'] == 6, 'Age']

5    30.726645
Name: Age, dtype: float64

In [12]:
titanic_dataframe.loc[titanic_dataframe['PassengerId'] == 20, 'Age']

19    27.915709
Name: Age, dtype: float64

In [13]:
titanic_dataframe.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

Так как не все классификаторы могут работать с категориальными признаками, в данном случае необходимо заменить их на числовые.

Также можно было бы выделить расширенные признаки, такие как статус пассажира (Mr., Mrs., Miss., Dr., Master. и т.д.), собственный индекс билета, индекс палубы и другие, но здесь ограничимся выбором более простых признаков.

### 1.5
В колонке Sex замените значения на 0, если пол "male", и на 1, если "female". В колонке Embarked замените параметры "U", "S", "C", "Q" на 0, 1, 2, 3 соответственно. Отбросьте колонки PassengerId, Name, Ticket, Cabin. Выведите первые 20 строк получившегося набора данных.

In [14]:
titanic_dataframe.loc[titanic_dataframe.Sex == 'male', 'Sex'] = 0
#titanic_dataframe.head(5)

In [15]:
titanic_dataframe.loc[titanic_dataframe.Sex == 'female', 'Sex'] = 1
#titanic_dataframe.head(5)

In [16]:
titanic_dataframe.loc[titanic_dataframe.Embarked == 'U', 'Embarked'] = 0
titanic_dataframe.loc[titanic_dataframe.Embarked == 'S', 'Embarked'] = 1
titanic_dataframe.loc[titanic_dataframe.Embarked == 'C', 'Embarked'] = 2
titanic_dataframe.loc[titanic_dataframe.Embarked == 'Q', 'Embarked'] = 3
titanic_dataframe = titanic_dataframe.iloc[:, [1,2,4,5,6,7,9,11]]
titanic_dataframe.head(20)

Survived  Pclass Sex        Age  SibSp  Parch     Fare Embarked
0          0       3   0  22.000000      1      0   7.2500        1
1          1       1   1  38.000000      1      0  71.2833        2
2          1       3   1  26.000000      0      0   7.9250        1
3          1       1   1  35.000000      1      0  53.1000        1
4          0       3   0  35.000000      0      0   8.0500        1
5          0       3   0  30.726645      0      0   8.4583        3
6          0       1   0  54.000000      0      0  51.8625        1
7          0       3   0   2.000000      3      1  21.0750        1
8          1       3   1  27.000000      0      2  11.1333        1
9          1       2   1  14.000000      1      0  30.0708        2
10         1       3   1   4.000000      1      1  16.7000        1
11         1       1   1  58.000000      0      0  26.5500        1
12         0       3   0  20.000000      0      0   8.0500        1
13         0       3   0  39.000000      1      5  31.2750        1
14         0       3   1  14.000000      0      0   7.8542        1
15         1       2   1  55.000000      0      0  16.0000        1
16         0       3   0   2.000000      4      1  29.1250        3
17         1       2   0  30.726645      0      0  13.0000        1
18         0       3   1  31.000000      1      0  18.0000        1
19         1       3   1  27.915709      0      0   7.2250        2

### 1.6
Приступим к построению классификаторов. **Условимся, что, если функция или объект модели имеют параметр random_state, то устанавливаем его равным 17 в каждом из случаев.**

Разделите данные на тренировочный и тестовый датасеты, установив размер тестового как 0.25 (первая колонка Survived является целевой, поэтому необходимо сначала ее отделить от признаков).

In [17]:
from sklearn.model_selection import train_test_split 

In [33]:
x = titanic_dataframe.iloc[:, 1:]
y = titanic_dataframe.iloc[:, 0]

In [34]:
y

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

In [35]:
x

Pclass Sex        Age  SibSp  Parch     Fare Embarked
0         3   0  22.000000      1      0   7.2500        1
1         1   1  38.000000      1      0  71.2833        2
2         3   1  26.000000      0      0   7.9250        1
3         1   1  35.000000      1      0  53.1000        1
4         3   0  35.000000      0      0   8.0500        1
..      ...  ..        ...    ...    ...      ...      ...
886       2   0  27.000000      0      0  13.0000        1
887       1   1  19.000000      0      0  30.0000        1
888       3   1  27.915709      1      2  23.4500        1
889       1   0  26.000000      0      0  30.0000        2
890       3   0  32.000000      0      0   7.7500        3

[891 rows x 7 columns]

In [36]:
x.shape

(891, 7)

In [37]:
y.shape

(891,)

In [38]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.25, random_state=17)

### 1.7
Обучите на полученных выборках несколько классификаторов, которые импортированы ниже.

**Примечание: в методе ближайших соседей используйте количество соседей, равное 5.**

In [39]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression

In [40]:
knn = KNeighborsClassifier(n_neighbors=5)
knn_model = knn.fit(x_train, y_train)
knn_pred = knn_model.predict(x_test)

In [41]:
gnb = GaussianNB()
gnb_model = gnb.fit(x_train, y_train)
gnb_pred = gnb_model.predict(x_test)

In [42]:
dct = DecisionTreeClassifier()
dct_model = dct.fit(x_train, y_train)
dct_pred = dct_model.predict(x_test)

In [43]:
lr = LogisticRegression()
lr_model = lr.fit(x_train, y_train)
lr_pred =lr_model.predict(x_test)

Очевидно, в этом случае использовать метрику accuracy, которую мы применяли до этого, не совсем правильно: она не даст верную оценку классификатору, потому как выживших явно меньше, чем погибших.

Вспомним про такие метрики как точность и полнота. Здесь они подойдут хорошо, так как у нас имеются положительный и отрицательный классы, и определить корректно один из них может оказаться более важным .

### 1.8
В каждом из пунктов следующего теста выберите, что важнее максимизировать: точность (precision) или полноту (recall). Выпишите ответы.

1. Вероятность того, что при определенной поломке самолета он сможет долететь до пункта назначения (1 - долетел, 0 - не долетел).
2. Предсказание, представляет ли человек опасность, по анализу психического состояния (1 - представляет опасность, 0 - не представляет опасности).
3. Предсказание ухода клиента (1 - клиент ушел, 0 - остался).
4. Выявление рака на основе медицинских показателей (1 - болен раком, 0 - здоров).
5. Предсказание летальности при наблюдаемой мутации (1 - выживание, 0 - летальный исход).
6. Определение важности происшествия для экстренных служб (1 - важно, 0 - неважно).
7. Окупятся ли вложения в бизнес (1 - окупятся, 0 - не окупятся).

Ваш ответ:

1. recall
2. recall
3. precision
4. recall
5. recall
6. recall
7. precision

### 1.9
Определите, что важнее использовать в случае датасета Titanic: полноту или точность. Объясните, почему.

### 1.10
Для каждого классификатора выведите матрицу ошибок и самостоятельно рассчитайте метрику recall (расчеты должны присутствовать). Проверьте расчеты, воспользовавшись встроенной метрикой recall_score.

1.9 
Зависит от целей исследования.
Если стоит вопрос о том, насколько точно модель предсказывает выжил человек или нет, то precision.
А если вопрос о том, насколько модель точно предсказывает, что человек действительно выжил из выборки выживших, предсказанной моделью - recall.

In [44]:
from sklearn.metrics import confusion_matrix, recall_score

In [45]:
knn_cm = confusion_matrix(y_test, knn_pred)
knn_cm

array([[113,  18],
       [ 46,  46]], dtype=int64)

46 экземпляров 0 класса путаются с 46 экземплярами 1 класса

In [48]:
from sklearn.metrics import recall_score
knn_recall = recall_score(y_test, knn_pred, average='micro')
knn_recall

0.7130044843049327

In [49]:
gnb_cm = confusion_matrix(y_test, gnb_pred)
gnb_cm

array([[111,  20],
       [ 27,  65]], dtype=int64)

27 экземпляров 0 класса путаются с 65 экземплярами 1 класса
20 экземпляров 1 класса путаются 111 с экземплярами 0 класса

In [50]:
gnb_recall = recall_score(y_test, gnb_pred, average='micro')
gnb_recall

0.7892376681614349

In [51]:
dct_cm = confusion_matrix(y_test, dct_pred)
dct_cm

array([[111,  20],
       [ 26,  66]], dtype=int64)

26 экземпляров 0 класса путаются с 66 экземплярами 1 класса
20 экземпляров 1 класса путаются 111 с экземплярами 0 класса

In [154]:
dct_recall = recall_score(y_test, dct_pred, average='micro')
dct_recall

1.0

In [52]:
lr_cm = confusion_matrix(y_test, lr_pred)
lr_cm

array([[113,  18],
       [ 28,  64]], dtype=int64)

28 экземпляров 0 класса путаются с 64 экземплярами 1 класса
18 экземпляров 1 класса путаются 113 с экземплярами 0 класса

In [155]:
lr_recall = recall_score(y_test, lr_pred, average='micro')
lr_recall

0.9820627802690582

### 1.11
Выберите ту модель, на которой метрика recall давала лучший ответ, и рассчитайте для нее precision, используя встроенную функцию.

In [53]:
from sklearn.metrics import precision_score

In [56]:
lr_precision = precision_score(y_test, lr_pred, average='micro')
lr_precision

0.7937219730941704

In [55]:
dct_precision = precision_score(y_test, dct_pred, average='micro')
dct_precision

0.7937219730941704